File: backend/src/services/user.service.ts
import bcrypt from 'bcryptjs';
import { Resend } from 'resend';
import * as db from '../db';
import config from '../config';
import AppError from '../utils/AppError';
import type { PaginatedResult } from '../types';

const CLUB_ROLES = ['club_admin', 'asset_manager', 'coach'];

// TODO: restore random generation before production
function generateTempPassword(): string {
  return '123456';
}

export async function listUsers(
  clubId: string,
  { role, is_active, page = 1, limit = 20 }: {
    role?: string;
    is_active?: string;
    page?: number | string;
    limit?: number | string;
  }
): Promise<PaginatedResult<Record<string, unknown>>> {
  const offset = (Number(page) - 1) * Number(limit);
  const conditions = ['club_id = $1'];
  const params: unknown[] = [clubId];

  if (role) conditions.push(`role = $${params.push(role)}`);
  if (is_active !== undefined) conditions.push(`is_active = $${params.push(is_active === 'true')}`);

  const where = conditions.join(' AND ');
  const [{ rows }, { rows: countRows }] = await Promise.all([
    db.query<Record<string, unknown>>(
      `SELECT id, name, email, phone, role, is_active, created_at
       FROM users WHERE ${where} ORDER BY name ASC
       LIMIT $${params.push(Number(limit))} OFFSET $${params.push(offset)}`,
      params
    ),
    db.query<{ count: string }>(`SELECT COUNT(*) FROM users WHERE ${where}`, params.slice(0, -2)),
  ]);
  return { data: rows, total: parseInt(countRows[0].count), page: Number(page), limit: Number(limit) };
}

export async function getUser(userId: string, clubId: string): Promise<Record<string, unknown>> {
  const { rows } = await db.query<Record<string, unknown>>(
    'SELECT id, name, email, phone, role, is_active, created_at FROM users WHERE id = $1 AND club_id = $2',
    [userId, clubId]
  );
  if (!rows.length) throw new AppError('User not found', 404);

  const { rows: teamRows } = await db.query<Record<string, unknown>>(
    `SELECT tm.team_id, tm.team_role, t.name AS team_name, t.gender, t.age_group
     FROM   team_members tm
     JOIN   teams t ON t.id = tm.team_id
     WHERE  tm.user_id = $1
     ORDER BY t.name ASC`,
    [userId]
  );

  return { ...rows[0], teams: teamRows };
}

export async function updateUser(
  targetId: string,
  clubId: string,
  { name, phone, role }: { name?: string; phone?: string; role?: string }
): Promise<Record<string, unknown>> {
  if (role && !CLUB_ROLES.includes(role)) throw new AppError('Invalid role', 400);

  if (role && role !== 'club_admin') {
    const { rows } = await db.query<{ count: string }>(
      `SELECT COUNT(*) FROM users
       WHERE club_id = $1 AND role = 'club_admin' AND is_active = true AND id != $2`,
      [clubId, targetId]
    );
    if (parseInt(rows[0].count) === 0) {
      throw new AppError('Cannot demote the last club admin', 409);
    }
  }

  const { rows } = await db.query<Record<string, unknown>>(
    `UPDATE users SET
       name  = COALESCE($1, name),
       phone = COALESCE($2, phone),
       role  = COALESCE($3::user_role, role)
     WHERE id = $4 AND club_id = $5
     RETURNING id, name, email, phone, role, is_active`,
    [name ?? null, phone ?? null, role ?? null, targetId, clubId]
  );
  if (!rows.length) throw new AppError('User not found', 404);
  return rows[0];
}

export async function deactivateUser(
  targetId: string,
  clubId: string,
  requesterId: string
): Promise<void> {
  if (targetId === requesterId) throw new AppError('You cannot deactivate your own account', 400);
  const { rows } = await db.query<{ id: string }>(
    'UPDATE users SET is_active = false WHERE id = $1 AND club_id = $2 RETURNING id',
    [targetId, clubId]
  );
  if (!rows.length) throw new AppError('User not found', 404);
}

export async function createUser(
  clubId: string,
  { email, name, role = 'coach', phone }: {
    email: string;
    name: string;
    role?: string;
    phone?: string;
  }
): Promise<Record<string, unknown>> {
  if (!email) throw new AppError('email is required', 400);
  if (!name?.trim()) throw new AppError('name is required', 400);
  if (!CLUB_ROLES.includes(role)) throw new AppError('Invalid role', 400);

  const { rows: existing } = await db.query<{ id: string }>(
    'SELECT id FROM users WHERE email = $1',
    [email.toLowerCase()]
  );
  if (existing.length) throw new AppError('This email is already registered', 409);

  const tempPassword = generateTempPassword();
  const passwordHash = await bcrypt.hash(tempPassword, 10);

  const { rows } = await db.query<Record<string, unknown>>(
    `INSERT INTO users (email, password_hash, name, phone, club_id, role, email_verified)
     VALUES ($1, $2, $3, $4, $5, $6, true)
     RETURNING id, email, name, phone, club_id, role, is_active, created_at`,
    [email.toLowerCase(), passwordHash, name.trim(), phone ?? null, clubId, role]
  );

  new Resend(config.resend.apiKey).emails.send({
    from: config.resend.fromEmail,
    to: email,
    subject: 'Welcome to SportStock',
    text: `Hi ${name},\n\nYour account has been created.\nEmail: ${email}\nTemporary password: ${tempPassword}\n\nPlease log in and change your password immediately.`,
  }).catch(() => {});

  return rows[0];
}

File: api/src/SportStock.Api/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using SportStock.Api.Data;
using SportStock.Api.Data.Entities;
using SportStock.Api.Data.Enums;
using SportStock.Api.Dtos.Common;
using SportStock.Api.Dtos.Users;
using SportStock.Api.Exceptions;
using SportStock.Api.Integrations;

namespace SportStock.Api.Services;

// V2 rewrite: users no longer carry ClubId or Role directly — membership is
// managed through the club_memberships table. All club-scoped user operations
// now join through ClubMembership.
//
// Preserved behaviours:
//   - Temp password hardcoded to "123456" until Resend wiring is restored.
//   - "Last active club_admin" demote is rejected with 409.
//   - super_admin is created out-of-band and is not returned here.
internal sealed class UserService(
    SportStockDbContext db,
    ILogger<UserService> log,
    ISupabaseStorage storage) : IUserService
{
    private static readonly ClubRole[] ClubRoles =
    {
        ClubRole.ClubAdmin,
        ClubRole.AssetManager,
        ClubRole.Coach,
    };

    // TODO: restore real temp-password generation before production
    private static string GenerateTempPassword() => "123456";

    public async Task<PaginatedResult<UserListItem>> ListAsync(
        Guid clubId,
        string? roleFilter,
        bool? isActiveFilter,
        int page,
        int limit,
        CancellationToken ct = default)
    {
        var safePage = page < 1 ? 1 : page;
        var safeLimit = limit < 1 ? 20 : limit;
        var offset = (safePage - 1) * safeLimit;

        var query = db.ClubMemberships
            .IgnoreQueryFilters()
            .Where(m => m.ClubId == clubId);

        if (TryParseRole(roleFilter, out var role))
            query = query.Where(m => m.Role == role);
        if (isActiveFilter is not null)
            query = query.Where(m => m.IsActive == isActiveFilter.Value);

        var total = await query.CountAsync(ct);

        var rows = await query
            .OrderBy(m => m.User.FirstName)
            .ThenBy(m => m.User.LastName)
            .Skip(offset)
            .Take(safeLimit)
            .Select(m => new UserListItem
            {
                Id = m.UserId,
                Name = m.User.FirstName + " " + m.User.LastName,
                Email = m.User.Email,
                Phone = m.User.Phone,
                Role = m.Role,
                IsActive = m.IsActive,
                CreatedAt = m.User.CreatedAt,
                AvatarUrl = m.User.AvatarUrl,
            })
            .ToListAsync(ct);

        return new PaginatedResult<UserListItem>
        {
            Data = rows,
            Total = total,
            Page = safePage,
            Limit = safeLimit,
        };
    }

    public async Task<UserDetailResponse> GetAsync(Guid userId, Guid clubId, CancellationToken ct = default)
    {
        var membership = await db.ClubMemberships
            .IgnoreQueryFilters()
            .Where(m => m.UserId == userId && m.ClubId == clubId)
            .Select(m => new UserDetailResponse
            {
                Id = m.UserId,
                Name = m.User.FirstName + " " + m.User.LastName,
                Email = m.User.Email,
                Phone = m.User.Phone,
                Role = m.Role,
                IsActive = m.IsActive,
                CreatedAt = m.User.CreatedAt,
                AvatarUrl = m.User.AvatarUrl,
            })
            .FirstOrDefaultAsync(ct);

        if (membership is null) throw new AppException("User not found", 404);

        membership.Teams = await db.TeamMembers
            .IgnoreQueryFilters()
            .Where(tm => tm.UserId == userId)
            .OrderBy(tm => tm.Team.Name)
            .Select(tm => new UserTeamMembership
            {
                TeamId = tm.TeamId,
                TeamRole = tm.TeamRole,
                TeamName = tm.Team.Name,
                Gender = tm.Team.Gender,
                AgeGroup = tm.Team.AgeGroup,
            })
            .ToListAsync(ct);

        return membership;
    }

    public async Task<UserListItem> CreateAsync(
        Guid clubId, CreateUserRequest req, CancellationToken ct = default)
    {
        var roleName = req.Role ?? "coach";
        if (!TryParseRole(roleName, out var role) || !ClubRoles.Contains(role))
            throw new AppException("Invalid role", 400);

        var emailNormalized = req.Email.Trim().ToLowerInvariant();

        // Check if user already exists (platform-wide)
        var existingUser = await db.Users
            .IgnoreQueryFilters()
            .FirstOrDefaultAsync(u => u.Email == emailNormalized, ct);

        User user;
        if (existingUser is not null)
        {
            // User already exists; just add them to this club if not already a member
            var alreadyMember = await db.ClubMemberships
                .IgnoreQueryFilters()
                .AnyAsync(m => m.UserId == existingUser.Id && m.ClubId == clubId, ct);
            if (alreadyMember)
                throw new AppException("This user is already a member of this club", 409);
            user = existingUser;
        }
        else
        {
            // Create new platform user
            var tempPassword = GenerateTempPassword();
            var passwordHash = BCrypt.Net.BCrypt.HashPassword(tempPassword, 10);

            // Parse name from req.Name (split on first space)
            var nameParts = (req.Name ?? "").Trim().Split(' ', 2, StringSplitOptions.RemoveEmptyEntries);
            var firstName = nameParts.Length > 0 ? nameParts[0] : req.Name?.Trim() ?? string.Empty;
            var lastName = nameParts.Length > 1 ? nameParts[1] : string.Empty;

            user = new User
            {
                Id = Guid.NewGuid(),
                Email = emailNormalized,
                PasswordHash = passwordHash,
                FirstName = firstName,
                LastName = lastName,
                Phone = req.Phone,
                EmailVerified = true,
                IsActive = true,
            };
            db.Users.Add(user);
            await db.SaveChangesAsync(ct);

            log.LogWarning(
                "EMAIL STUB (welcome): would send temp password {TempPassword} to {Email}",
                tempPassword, emailNormalized);
        }

        // Create club membership
        var membership = new ClubMembership
        {
            Id = Guid.NewGuid(),
            ClubId = clubId,
            UserId = user.Id,
            Role = role,
            IsActive = true,
            JoinedAt = DateTime.UtcNow,
        };
        db.ClubMemberships.Add(membership);
        await db.SaveChangesAsync(ct);

        return new UserListItem
        {
            Id = user.Id,
            Name = user.FirstName + " " + user.LastName,
            Email = user.Email,
            Phone = user.Phone,
            Role = role,
            IsActive = true,
            CreatedAt = user.CreatedAt,
            AvatarUrl = user.AvatarUrl,
        };
    }

    public async Task<UserListItem> UpdateAsync(
        Guid targetId, Guid clubId, UpdateUserRequest req, CancellationToken ct = default)
    {
        ClubRole? newRole = null;
        if (req.Role is not null)
        {
            if (!TryParseRole(req.Role, out var parsed) || !ClubRoles.Contains(parsed))
                throw new AppException("Invalid role", 400);
            newRole = parsed;
        }

        if (newRole is ClubRole r && r != ClubRole.ClubAdmin)
        {
            var otherActiveAdmins = await db.ClubMemberships
                .IgnoreQueryFilters()
                .CountAsync(m => m.ClubId == clubId
                                && m.Role == ClubRole.ClubAdmin
                                && m.IsActive
                                && m.UserId != targetId, ct);
            if (otherActiveAdmins == 0)
                throw new AppException("Cannot demote the last club admin", 409);
        }

        var membership = await db.ClubMemberships
            .IgnoreQueryFilters()
            .Include(m => m.User)
            .FirstOrDefaultAsync(m => m.UserId == targetId && m.ClubId == clubId, ct);
        if (membership is null) throw new AppException("User not found", 404);

        if (req.Name is not null)
        {
            var nameParts = req.Name.Trim().Split(' ', 2, StringSplitOptions.RemoveEmptyEntries);
            membership.User.FirstName = nameParts.Length > 0 ? nameParts[0] : req.Name.Trim();
            membership.User.LastName = nameParts.Length > 1 ? nameParts[1] : string.Empty;
        }
        if (req.Phone is not null) membership.User.Phone = req.Phone;
        if (newRole is not null) membership.Role = newRole.Value;

        await db.SaveChangesAsync(ct);

        return new UserListItem
        {
            Id = membership.UserId,
            Name = membership.User.FirstName + " " + membership.User.LastName,
            Email = membership.User.Email,
            Phone = membership.User.Phone,
            Role = membership.Role,
            IsActive = membership.IsActive,
            CreatedAt = membership.User.CreatedAt,
            AvatarUrl = membership.User.AvatarUrl,
        };
    }

    public async Task DeactivateAsync(
        Guid targetId, Guid clubId, Guid requesterId, CancellationToken ct = default)
    {
        if (targetId == requesterId)
            throw new AppException("You cannot deactivate your own account", 400);

        var rows = await db.ClubMemberships
            .IgnoreQueryFilters()
            .Where(m => m.UserId == targetId && m.ClubId == clubId)
            .ExecuteUpdateAsync(s => s.SetProperty(m => m.IsActive, false), ct);
        if (rows == 0) throw new AppException("User not found", 404);
    }

    public async Task<UploadAvatarResponse> UploadAvatarAsync(
        Guid userId, Guid clubId, Stream content,
        string contentType, string fileName, CancellationToken ct = default)
    {
        var ext = Path.GetExtension(fileName).TrimStart('.');
        var path = $"avatars/{clubId}/{userId}_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}.{ext}";
        var url = await storage.UploadAsync(path, content, contentType, ct);

        await db.Users
            .Where(u => u.Id == userId)
            .ExecuteUpdateAsync(s => s
                .SetProperty(u => u.AvatarUrl, url)
                .SetProperty(u => u.UpdatedAt, DateTime.UtcNow), ct);

        return new UploadAvatarResponse { AvatarUrl = url };
    }

    private static bool TryParseRole(string? value, out ClubRole role)
    {
        switch (value)
        {
            case "club_admin":    role = ClubRole.ClubAdmin;    return true;
            case "asset_manager": role = ClubRole.AssetManager; return true;
            case "coach":         role = ClubRole.Coach;        return true;
            case "accountant":    role = ClubRole.Accountant;   return true;
            default:              role = default;               return false;
        }
    }
}
